File: Longest_Valid_Parentheses/sol2.cpp
/* -.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.
* File Name : sol2.cpp
* Purpose : Stack solution
* Creation Date : 28-01-2014
* Last Modified : Tue Jan 28 22:04:48 2014
* Created By : wdd 
_._._._._._._._._._._._._._._._._._._._._.*/
#include "general.h"
    int longestValidParentheses(string s) {
        vector<pair<char, int>> sta; //sta is to record the symbol and its index
        int maxP = 0;
        for (int i = 0; i < s.length(); ++ i){
            char cur = s[i];
            if(cur == '(') sta.push_back(make_pair(cur, i));// if '(', just push
            else{
                if(!sta.empty() && sta.back().first == '('){// a valid parentheses formed
                    sta.pop_back();
                    maxP = max(maxP, i-(sta.empty()?-1:sta.back().second));// if sta is empty, the full string matches and the length is the current idx
                }
                else
                    sta.push_back(make_pair(cur, i));
            }
        }
        return maxP;
    }
int main(int argc, const char* argv[])
{
	return 0;
}

File: Longest_Valid_Parentheses/sol2.py
#!/usr/bin/python
# --------------------------------------- 
# File Name : sol2.py
# Creation Date : 25-11-2014
# Last Modified : Tue Nov 25 20:22:38 2014
# Created By : wdd 
# --------------------------------------- 
import sys, re, string, pickle
from itertools import *
class Solution:
    # @param s, a string
    # @return an integer
    def longestValidParentheses(self, s):
        sta, max_len = [], 0
        for i, val in enumerate(s):
            if not len(sta) or val == '(' or sta[-1][0] == ')':
                sta += [(val, i)]
            else:
                sta.pop()
                tmp_len = i - (sta[-1][1] if len(sta) else -1)
                if tmp_len > max_len:
                    max_len = tmp_len
        return max_len
if __name__ == "__main__": main()
